controllers/kubeadmconfig_controller_test.go

```diff
@@ -554,6 +561,180 @@ func TestReconcileIfJoinNodesAndControlPlaneIsReady(t *testing.T) {
+						},
+					},
+				},
+			},
+			validateDiscovery: func(t *testing.T, c *cabpkV1alpha2.KubeadmConfig) {
+				d := c.Spec.JoinConfiguration.Discovery
+				if d.BootstrapToken != nil {
+					t.Fatal("BootstrapToken should not be created when DiscoveryFile is defined")
+				}
+			},
+		},
+		{
+			name:    "Respect discoveryConfiguration.BootstrapToken.APIServerEndpoint",
+			cluster: goodcluster,
+			config: &cabpkV1alpha2.KubeadmConfig{
+				Spec: cabpkV1alpha2.KubeadmConfigSpec{
+					JoinConfiguration: &kubeadmv1beta1.JoinConfiguration{
+						Discovery: kubeadmv1beta1.Discovery{
+							BootstrapToken: &kubeadmv1beta1.BootstrapTokenDiscovery{
+								APIServerEndpoint: "https://bar.com:6443",
+							},
+						},
+					},
+				},
+			},
+			validateDiscovery: func(t *testing.T, c *cabpkV1alpha2.KubeadmConfig) {
+				d := c.Spec.JoinConfiguration.Discovery
+				if d.BootstrapToken.APIServerEndpoint != "https://bar.com:6443" {
+					t.Fatalf("BootstrapToken.APIServerEndpoint=https://bar.com:6443 expected, got %s", d.BootstrapToken.APIServerEndpoint)
+				}
+			},
+		},
+		{
+			name:    "Respect discoveryConfiguration.BootstrapToken.Token",
+			cluster: goodcluster,
+			config: &cabpkV1alpha2.KubeadmConfig{
+				Spec: cabpkV1alpha2.KubeadmConfigSpec{
+					JoinConfiguration: &kubeadmv1beta1.JoinConfiguration{
+						Discovery: kubeadmv1beta1.Discovery{
+							BootstrapToken: &kubeadmv1beta1.BootstrapTokenDiscovery{
+								Token: "abcdef.0123456789abcdef",
+							},
+						},
+					},
+				},
+			},
+			validateDiscovery: func(t *testing.T, c *cabpkV1alpha2.KubeadmConfig) {
+				d := c.Spec.JoinConfiguration.Discovery
+				if d.BootstrapToken.Token != "abcdef.0123456789abcdef" {
+					t.Fatalf("BootstrapToken.Token=abcdef.0123456789abcdef expected, got %s", d.BootstrapToken.Token)
+				}
+			},
+		},
+		{
+			name:    "Respect discoveryConfiguration.BootstrapToken.CACertHashes",
+			cluster: goodcluster,
+			config: &cabpkV1alpha2.KubeadmConfig{
+				Spec: cabpkV1alpha2.KubeadmConfigSpec{
+					JoinConfiguration: &kubeadmv1beta1.JoinConfiguration{
+						Discovery: kubeadmv1beta1.Discovery{
+							BootstrapToken: &kubeadmv1beta1.BootstrapTokenDiscovery{
+								CACertHashes: dummyCAHash,
+							},
+						},
+					},
+				},
+			},
+			validateDiscovery: func(t *testing.T, c *cabpkV1alpha2.KubeadmConfig) {
+				d := c.Spec.JoinConfiguration.Discovery
+				if !reflect.DeepEqual(d.BootstrapToken.CACertHashes, dummyCAHash) {
+					t.Fatalf("BootstrapToken.CACertHashes=%s expected, got %s", dummyCAHash, d.BootstrapToken.Token)
+				}
+			},
+		},
+	}
+
+	for _, rt := range useCases {
+		t.Run(rt.name, func(t *testing.T) {
+			err := k.reconcileDiscovery(rt.cluster, rt.config)
+
+			if err == nil && rt.expectedError {
+				t.Error("expected error, got nil")
+			}
+			if err != nil && !rt.expectedError {
+				t.Errorf("expected nil, got error %v", err)
+			}
+
+			if rt.validateDiscovery != nil {
+				rt.validateDiscovery(t, rt.config)
```

Comment:
Unless this function panics (it absolutely should not) there is no assertion here or the error is being ignored.